### PR TITLE
Fix a race between ApplySnapshot and splitQueue.

### LINKED ISSUE #3164: race: ApplySnapshot vs (*splitQueue).shouldQueue()
the racing data is  `r.stats, err = newRangeStats(desc.RangeID, batch)`.

Got this locally on my branch, but I don't think it's related to my changes. Unfortunately, don't see which test triggered this (but it's in storage and uses `readIntFromEngines`, so it's probably one of the replication tests.

```
make testrace PKG=./storage TESTS=. 2>&1 | less
...
==================
WARNING: DATA RACE
Write by goroutine 21:
  github.com/cockroachdb/cockroach/storage.(*Replica).ApplySnapshot()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/replica_raftstorage.go:503 +
0xa15
  github.com/cockroachdb/cockroach/multiraft.(*writeTask).start.func1()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/multiraft/storage.go:236 +0xbe6
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:88 +0x5f

Previous read by goroutine 179:
  github.com/cockroachdb/cockroach/storage.(*splitQueue).shouldQueue()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:83 +0x2fd
  github.com/cockroachdb/cockroach/storage.(*baseQueue).MaybeAdd()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/queue.go:210 +0x452
  github.com/cockroachdb/cockroach/storage.(*Store).systemGossipUpdate()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/store.go:648 +0x2c6
  github.com/cockroachdb/cockroach/storage.(*Store).(github.com/cockroachdb/cockroach/storage.systemG
ossipUpdate)-fm()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/storage/store.go:519 +0x37

Goroutine 21 (running) created at:
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x6f
  github.com/cockroachdb/cockroach/multiraft.(*writeTask).start()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/multiraft/storage.go:257 +0xcd
  github.com/cockroachdb/cockroach/multiraft.(*state).start.func1()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:559 +0x222
  github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:88 +0x5f

Goroutine 179 (finished) created at:
  github.com/cockroachdb/cockroach/gossip.(*Gossip).updateSystemConfig()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:390 +0x3c0
  github.com/cockroachdb/cockroach/gossip.(*Gossip).(github.com/cockroachdb/cockroach/gossip.updateSy
stemConfig)-fm()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:158 +0x81
  github.com/cockroachdb/cockroach/gossip.(*infoStore).processCallbacks.func1()
      /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/gossip/infostore.go:242 +0xe5
==================
```

### PR DESCRIPTION
**Pull request by bdarnell**
Fixes #3164.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3269)

### CODE REVIEW

**Comment by BramGruneir** [comment]
LGTM  Nice catch.

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3269)*

File: storage/stats.go [hunk 1]
```diff
@@ -51,6 +51,13 @@ func newRangeStats(rangeID roachpb.RangeID, e engine.Engine) (*rangeStats, error
 	return rs, nil
 }
 
+// Replace *rs with the contents of *other.
+func (rs *rangeStats) Replace(other *rangeStats) {
+	rs.Lock()
+	defer rs.Unlock()
+	rs.MVCCStats = other.MVCCStats
```
Comment (tschottdorf) [inline storage/stats.go]: should you technically be lock-defer-unlock `other` as well?

**Comment by tschottdorf** [comment]
what's the exact race here? Could you add a brief explanation to the commit message?

**Comment by tschottdorf** [comment]
ah, didn't think of the Mutex. I take it back.

**Comment by bdarnell** [comment]
The race is in the assignment to `r.stats`. `rangeStats` has an internal mutex so `r.stats` is accessed without synchronization. The fix is to overwrite the contents of the stats object in place.

File: storage/stats.go (hunk 1 above)
Comment (bdarnell) [inline storage/stats.go]: Yeah, I guess so. There's a theoretical deadlock risk if you lock both of them but we don't need to worry about it.

Comment (tschottdorf) [inline storage/stats.go]: LGTM

**Pull request merged**